dhnx/simulation.py

```diff
@@ -10,23 +10,440 @@
+            m : pd.DataFrame
+                DataFrame with all know mass flow of
+                consumers and producer.
+            """
+            producers = [name for name in m.columns if re.search('producers', name)]
+
+            assert len(producers) == 1, "Currently, only one producer allowed."
+
+            m.loc[:, producers] = - m.loc[:, ~m.columns.isin(producers)].sum(1)
+
+            return m
+
+        self.inc_mat = nx.incidence_matrix(self.nx_graph, oriented=True).todense()
+
+        mass_flow = pd.DataFrame(
+            0,
+            columns=self.nx_graph.nodes(),
+            index=self.thermal_network.timeindex
+        )
+
+        input_data = self._concat_sequences('mass_flow')
+
+        mass_flow.loc[:, input_data.columns] = input_data
+
+        self.input_data.mass_flow = _set_producers_mass_flow(mass_flow)
+
+    def _solve_hydraulic_eqn(self, tolerance=1e-10):
+
+        pipes_mass_flow = {}
+
+        for t in self.thermal_network.timeindex:
+
+            x, residuals, rank, s = np.linalg.lstsq(
+                self.inc_mat,
+                self.input_data.mass_flow.loc[t, :],
+                rcond=None
+            )
+
+            assert residuals < tolerance,\
+                f"Residuals {residuals} are larger than tolerance {tolerance}!"
+
+            pipes_mass_flow.update({t: x})
+
+        self.results['pipes-mass_flow'] = pd.DataFrame.from_dict(
+            pipes_mass_flow,
+            orient='index',
+            columns=self.nx_graph.edges()
+        )
+
+        # TODO: Calculate these
+        # 'pipes-pressure_losses'
+        # NOTE: pipes have distributed and localized pressure losses zeta_dis * L/D**5
+        # nodes-pressure_losses
+        # NOTE: nodes have only localized pressure losses
+        # 'global-pressure_losses'
+        # 'producers-pump_power'
+
+        def _calculate_reynolds():
+            r"""
+            Calculates the reynolds number.
+
+            ..math::
+
+                Re = \frac{4\dot{m}{\pi\mu D}
+
+            Returns
+            -------
+
+            re : pd.DataFrame
+            """
+            pipes_mass_flow = self.results['pipes-mass_flow']
+
+            diameter = self.thermal_network.components.pipes[['from_node', 'to_node', 'diameter_mm']]
+
+            diameter = 1e-3 * diameter.set_index(['from_node', 'to_node'])['diameter_mm']
+
+            re = 4 * pipes_mass_flow.divide(diameter, axis='columns') \
+                / (np.pi * self.mu)
+
+            return re
+
+        def _calculate_lambda(re):
+            r"""
+            Calculates the darcy friction factor.
+
+            ..math::
+
+                \lambda = 0.007 \cdot Re^{-0.13} \cdot D ^ {-0.14}
+
+            Parameters
+            ----------
+            re
+
+            Returns
+            -------
+
+            lamb : pd.DataFrame
+            """
+
+            factor_diameter = 1e-3 * self.thermal_network.components.pipes['diameter_mm'] ** -0.14
```

Comment:
```suggestion
            factor_diameter = (1e-3 * self.thermal_network.components.pipes['diameter_mm']) ** -0.14
```

Parenthesis are missing around diameter and its unit conversion (1e-3).

Reply:
This should not make a difference.

Reply:
Fixed with 4afe05d.